### DISCUSSION TITLE
三体

### DISCUSSION THREAD

**Discussion by franc2es** [creatorsand/.github#40] [Ideas]
玩家可选种地、军事和科技来升级点数，白天促进种地，闪电促进科技，夜晚降低军事，两玩家相邻时交战触发天气，科技点数高的控制此地盘天气，军事点数高的占领土地，种地后土地被占领可使占领者点数下降，地图上可拥有两名玩家，获胜规则为所有土地被占满后点数乘土地数最多者胜。

**Comment by franc2es** [comment]
科技、军事单独记，和地图有关联的就只有交战

**Comment by metawyc** [comment]
`// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract SurvivalGame {
    address public owner;
    uint256 public lastLightningTimestamp;
    uint256 public lastDroughtTimestamp;
    bool public isDaytime;
    uint256 public territorySize;
    uint256 public industrialPower;
    uint256 public technologyLevel;
    uint256 public militaryStrength;
    address public currentTerritoryOwner;

    constructor() {
        owner = msg.sender;
        lastLightningTimestamp = block.timestamp;
        lastDroughtTimestamp = block.timestamp;
        isDaytime = true;
        territorySize = 1;
        industrialPower = 0;
        technologyLevel = 0;
        militaryStrength = 0;
        currentTerritoryOwner = address(0);
    }

    modifier onlyOwner() {
        require(msg.sender == owner, "Only the owner can call this function");
        _;
    }

    function changeTerritoryOwner(address newOwner) external onlyOwner {
        currentTerritoryOwner = newOwner;
    }

    function lightningStrike() external onlyOwner {
        require(block.timestamp > lastLightningTimestamp + 1 days, "You can't trigger lightning yet.");
        industrialPower += 10;
        lastLightningTimestamp = block.timestamp;
    }

    function drought() external onlyOwner {
        require(block.timestamp > lastDroughtTimestamp + 1 days, "You can't trigger drought yet.");
        territorySize -= 1;
        technologyLevel += 10;
        lastDroughtTimestamp = block.timestamp;
    }

    function endDaytime() external onlyOwner {
        isDaytime = false;
        territorySize += 2;
    }

    function startDaytime() external onlyOwner {
        isDaytime = true;
        militaryStrength += 5;
    }

    function invadeTerritory(address target) external {
        require(militaryStrength > 50, "You need a military strength of at least 50 to invade.");
        currentTerritoryOwner = target;
    }

    function claimReward() external {
        require(currentTerritoryO` 初步的代码实现，尚未利用MUD

**Reply by franc2es** [reply]
contract MapSystem is System {
  function spawn(uint32 x, uint32 y) public {
    bytes32 player = addressToEntityKey(address(_msgSender()));
    require(!Player.get(player), "already spawned");

    // 获取地图配置信息
    (uint32 width, uint32 height, bytes memory terrain, bytes8[] memory owners) = MapConfig.get();

    bytes32 position = positionToEntityKey(x, y);
    require(!Obstruction.get(position), "this space is obstructed");

    // 确定当前玩家是玩家0还是玩家1（交替行动）
    uint8 currentPlayer = Player.get(player) ? 1 : 0;

    // 检查所选位置是否已经被占领
    require(owners[x + y * width] == bytes8(address(0)), "this space is already occupied");

    // 将当前玩家设置为领土占领者
    owners[x + y * width] = owners[currentPlayer];

    // 设置玩家已经生成并占领位置
    Player.set(player, true);
    Position.set(player, x, y);
    Movable.set(player, true);

    // 更新地图配置
    MapConfig.set(world, width, height, terrain, owners);
  }

  function move(uint32 x, uint32 y) public {
    bytes32 player = addressToEntityKey(_msgSender());
    require(Movable.get(player), "cannot move");

    // 获取地图配置信息
    (uint32 width, uint32 height, bytes memory terrain, bytes8[] memory owners) = MapConfig.get();

    bytes32 position = positionToEntityKey(x, y);
    require(!Obstruction.get(position), "this space is obstructed");

    // 确定当前玩家是玩家0还是玩家1
    uint32 currentPlayer = Player.get(player) ? 1 : 0;

    // 计算目标玩家是另一个玩家
    uint32 targetPlayer = (currentPlayer + 1) % 2;

    // 检查是否发生战斗，即检查所选位置是否属于另一个玩家
    require(owners[x + y * width] == owners[targetPlayer], "this space does not belong to the other player");

    // 更新玩家位置
    Position.set(player, x, y);
  }
}

**Comment by metawyc** [comment]
欢迎朋友们一起组队！！！可以加vx：wxid_ezxfl6plgnrv12或者metawyc

**Comment by franc2es** [comment]
pragma solidity ^0.8.0;

import { System } from "@latticexyz/world/src/System.sol";
import { Table, SchemaType, TableData } from "@latticexyz/world/src/Table.sol";
import { WorldContext } from "@latticexyz/world/src/World.sol";

contract LandConquestGame is System {
    Table private landTable;
    Table private playerTable;

    constructor(WorldContext ctx) System(ctx) {
        landTable = createTable("LandTable", SchemaType.Uint32);
        playerTable = createTable("PlayerTable", SchemaType.Uint32);
    }

    function joinGame() public {
        require(!playerTable.contains(msg.sender), "Player already in the game");
        playerTable.insert(msg.sender, 0); // 初始化玩家分数为0
    }

    function startDaytime() public {
        require(isRoot() && isDaytime(), "Only the root user can start daytime");
        // 实现白天逻辑
    }

    function startNighttime() public {
        require(isRoot() && isNighttime(), "Only the root user can start nighttime");
        // 实现夜晚逻辑
    }

    function activateLightning() public {
        require(isRoot() && isDaytime(), "Only the root user can activate lightning during the day");
        // 实现闪电逻辑
    }

    function deactivateLightning() public {
        require(isRoot() && isLightningActive(), "Only the root user can deactivate lightning during the day");
        // 实现关闭闪电逻辑
    }

    function increaseScore(uint256 score) public {
        require(isRoot() && isDaytime(), "Only the root user can increase score during the day");
        require(score > 0, "Score must be greater than 0");
        playerTable.set(msg.sender, playerTable.get(msg.sender) + score);
    }

    function triggerWar() public {
        require(isRoot() && isNighttime(), "Only the root user can trigger war at night");
        // 实现触发战争逻辑
    }

    function conquerLand() public {
        require(isRoot() && isWarActive(), "Only the root user can conquer land during war");
        // 实现占领土地逻辑
    }

    function toggleWeatherControl(bool controlWeather) public {
        require(isRoot() && isDaytime(), "Only the root user can toggle weather control during the day");
        // 实现控制天气逻辑
    }

    function endWar() public {
        require(isRoot() && isWarActive(), "Only the root user can end war");
        // 实现结束战争逻辑
    }

    function claimVictory() public {
        require(isRoot(), "Only the root user can claim victory");
        // 实现宣布胜利逻辑
    }
}改了点

**Comment by franc2es** [comment]
pragma solidity ^0.8.0;

contract SurvivalGame {
    address public owner;
    bool public isDaytime;
    uint256 public techPoints;
    uint256 public militaryPoints;

    // 阈值，根据你的游戏规则设置
    uint256 public techThreshold = 100;
    uint256 public militaryThreshold = 50;

    // 地图系统合约
    MapSystem public mapSystem;

    constructor(address _mapSystemAddress) {
        owner = msg.sender;
        isDaytime = true;
        techPoints = 0;
        militaryPoints = 0;
        mapSystem = MapSystem(_mapSystemAddress); // 初始化地图系统合约实例
    }

    modifier onlyOwner() {
        require(msg.sender == owner, "Only the owner can call this function");
        _;
    }

    modifier isNighttime() {
        require(!isDaytime, "It's not nighttime");
        _;
    }

    function changeWeather() internal {
        // 实现改变下一轮天气的逻辑
    }

    function conquerLandFromOpponent(address opponent) internal {
        // 实现获得对方土地的逻辑
    }

    // 获取科技点数
    function getTechnologyPoints() public view returns (uint256) {
        return techPoints;
    }

    // 获取军事点数
    function getMilitaryPoints() public view returns (uint256) {
        return militaryPoints;
    }

    // 触发战争逻辑
    function triggerWar() public onlyOwner isNighttime {
        uint256 techPoints = getTechnologyPoints();
        uint256 militaryPoints = getMilitaryPoints();

        if (techPoints > techThreshold) {
            // 科技点数高，改变下一轮的天气
            changeWeather();
        }

        if (militaryPoints > militaryThreshold) {
            // 军事点数高，触发战斗
            mapSystem.triggerBattle();
        }
    }

    // 开始白天
    function startDaytime() external onlyOwner {
        isDaytime = true;
    }

    // 开始夜晚
    function startNighttime() external onlyOwner {
        isDaytime = false;
    }

    // 增加科技点数
    function increaseTechPoints(uint256 points) external onlyOwner {
        techPoints += points;
    }

    // 增加军事点数
    function increaseMilitaryPoints(uint256 points) external onlyOwner {
        militaryPoints += points;
    }
}


   contract MapSystem is System {
function spawn(uint32 x, uint32 y) public {
bytes32 player = addressToEntityKey(address(_msgSender()));
require(!Player.get(player), "already spawned");

// 获取地图配置信息
(uint32 width, uint32 height, bytes memory terrain, bytes8[] memory owners) = MapConfig.get();

bytes32 position = positionToEntityKey(x, y);
require(!Obstruction.get(position), "this space is obstructed");

// 确定当前玩家是玩家0还是玩家1（交替行动）
uint8 currentPlayer = Player.get(player) ? 1 : 0;

// 检查所选位置是否已经被占领
require(owners[x + y * width] == bytes8(address(0)), "this space is already occupied");

// 将当前玩家设置为领土占领者
owners[x + y * width] = owners[currentPlayer];

// 设置玩家已经生成并占领位置
Player.set(player, true);
Position.set(player, x, y);
Movable.set(player, true);

// 更新地图配置
MapConfig.set(world, width, height, terrain, owners);
}

function move(uint32 x, uint32 y) public {
bytes32 player = addressToEntityKey(_msgSender());
require(Movable.get(player), "cannot move");

// 获取地图配置信息
(uint32 width, uint32 height, bytes memory terrain, bytes8[] memory owners) = MapConfig.get();

bytes32 position = positionToEntityKey(x, y);
require(!Obstruction.get(position), "this space is obstructed");

// 确定当前玩家是玩家0还是玩家1
uint32 currentPlayer = Player.get(player) ? 1 : 0;

// 计算目标玩家是另一个玩家
uint32 targetPlayer = (currentPlayer + 1) % 2;

// 检查是否发生战斗，即检查所选位置是否属于另一个玩家
require(owners[x + y * width] == owners[targetPlayer], "this space does not belong to the other player");

// 更新玩家位置
Position.set(player, x, y);
}
}

contract MapSystem {
    // 定义玩家结构体来存储玩家的状态
    struct Player {
        bool spawned;
        bool alive;
        uint32 x;
        uint32 y;
    }

    // 映射地址到玩家
    mapping(address => Player) public players;

    // 更改玩家的状态为已生成
    function spawn(uint32 x, uint32 y) public {
        require(!players[msg.sender].spawned, "Player already spawned");

        // 设置玩家的状态
        players[msg.sender] = Player(true, true, x, y);

        // 其他生成逻辑...
    }

    // 玩家移动到新位置
    function move(uint32 x, uint32 y) public {
        Player storage player = players[msg.sender];
        require(player.spawned, "Player has not spawned");
        require(player.alive, "Player is not alive");

        // 实现移动逻辑...

        // 检查是否与对手相邻
        address opponent = getAdjacentPlayer(x, y);
        if (opponent != address(0)) {
            // 触发战争
            triggerBattle(msg.sender, opponent);
        }
    }

    // 获取相邻的对手玩家
    function getAdjacentPlayer(uint32 x, uint32 y) internal view returns (address) {
        Player storage currentPlayer = players[msg.sender];
        uint32 currentPlayerX = currentPlayer.x;
        uint32 currentPlayerY = currentPlayer.y;

        // 检查上下左右相邻的位置
        if ((x == currentPlayerX && (y == currentPlayerY - 1 || y == currentPlayerY + 1)) ||
            (y == currentPlayerY && (x == currentPlayerX - 1 || x == currentPlayerX + 1))) {
            // 返回相邻对手玩家的地址
            address opponent = getOwner(x, y);
            if (opponent != address(0) && opponent != msg.sender) {
                return opponent;
            }
        }

        return address(0); // 没有相邻的对手玩家
    }

    // 获取指定位置的领地所有者
    function getOwner(uint32 x, uint32 y) internal view returns (address) {
        // 实现获取领地所有者的逻辑...
    }
//调用triggerWar
    
        }
    }
}


    
    }
}

**Comment by Demian101** [comment]
wow, looks fancy
